R/SCE-reducedDims.R

```diff
@@ -53,25 +53,118 @@ setReplaceMethod("reducedDimNames", c("SingleCellExperiment", "character"), func
+        return(NULL)
+        # To deprecate NULL and throw an error instead, remove the two lines above.
+        stop(
+            "'reducedDim(<", class(x), ">, ...) ",
+            "length(reducedDims(<", class(x), ">)) is 0'")
+    }
+
+    reducedDim(x, 1L, withDimnames)
+})
+
+#' @export
+setMethod("reducedDim", c("SingleCellExperiment", "numeric"), function(x, type=1, withDimnames=TRUE) {
+    x <- updateObject(x)
+    internals <- int_colData(x)[[.red_key]]
+
+    out <- tryCatch({
+        internals[, type]
+    }, error=function(err) {
+        .Deprecated(msg="NULL is deprecated.")
+        return(NULL)
+        # To deprecate NULL and throw an error instead, remove the two lines above.
+        stop("'reducedDim(<", class(x), ">, type=\"numeric\", ...)' ",
+             "invalid subscript 'type'\n", conditionMessage(err))
+    })
+
+    if (withDimnames) {
+        rownames(out) <- colnames(x)
+    }
+
+    out
+})
+
+#' @export
+setMethod("reducedDim", c("SingleCellExperiment", "character"), function(x, type, withDimnames=TRUE) {
+    msg <- paste0("'reducedDim(<", class(x), ">, type=\"character\", ...)' ",
+                  "invalid subscript 'type'")
 
+    x <- updateObject(x)
     internals <- int_colData(x)[[.red_key]]
-    out <- internals[,type]
-    if (!is.null(out) && withDimnames) {
+
+    out <- tryCatch({
+        internals[, type]
+    }, error=function(err) {
+        .Deprecated(msg="NULL is deprecated.")
+        return(NULL)
+        # To deprecate NULL and throw an error instead, remove the two lines above.
+        stop(msg, "\n'", type, "' not in names(reducedDims(<", class(x), ">))")
+    })
+
+    if (withDimnames) {
         rownames(out) <- colnames(x)
     }
+
     out
 })
 
 #' @export
-setReplaceMethod("reducedDim", "SingleCellExperiment", function(x, type=1, ..., value) {
+setReplaceMethod("reducedDim", c("SingleCellExperiment", "missing"), function(x, type, ..., value) {
+    type <- 1L
+
+    if (0L == length(reducedDimNames(x))){
+        ## Implementation 1
+        # https://github.com/drisso/SingleCellExperiment/pull/35#issuecomment-522258649
+        # type <- "unnamed"
+        ## Implementation 2
+        # `reducedDims<-`` above creates character(n) names
+        type <- character(1)
+        ## Implementation 3
+        # This would implement the same behaviour as SummarizedExperiment::assay<-
+        # stop("'reducedDim(<", class(x), ">) <- value' ", "length(reducedDims(<",
+        #      class(x), ">)) is 0")
+    }
+
+    reducedDim(x, type) <- value
+    x
+})
+
+#' @export
+setReplaceMethod("reducedDim", c("SingleCellExperiment", "numeric"), function(x, type = 1, ..., value) {
+    x <- updateObject(x)
+
+    internals <- int_colData(x)
+
+    if (!is.null(value) && !identical(nrow(value), ncol(x))) {
+        stop("replacement 'reducedDim' has a different number of rows than 'ncol(x)'")
+    }
+
+    # How do we want to deal with length(type) > 1 ?
```

Comment:
Don't worry about it, `[[` will fail soon enough.

Reply:
OK, but still, to avoid issues with the `if` on the next line, I need to test a single scalar.
Anyway, I'll remove the comment and test the first value. No one should ever notice anything unless they explicitly try to provide a vector of length > 1 ...

Reply:
Well, people will just get a warning, followed by an error, so it's no big deal.